core/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java

```diff
@@ -179,44 +182,48 @@ public Explanation explain(int topLevelDocId, SearchContext context, RescoreSear
         }
     }
 
-    @Override
-    public RescoreSearchContext parse(XContentParser parser, SearchContext context) throws IOException {
-        Token token;
-        String fieldName = null;
-        QueryRescoreContext rescoreContext = new QueryRescoreContext(this);
-        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
-            if (token == XContentParser.Token.FIELD_NAME) {
-                fieldName = parser.currentName();
-                if ("rescore_query".equals(fieldName)) {
-                    ParsedQuery parsedQuery = context.queryParserService().parse(parser);
-                    rescoreContext.setParsedQuery(parsedQuery);
-                }
-            } else if (token.isValue()) {
-                if ("query_weight".equals(fieldName)) {
-                    rescoreContext.setQueryWeight(parser.floatValue());
-                } else if ("rescore_query_weight".equals(fieldName)) {
-                    rescoreContext.setRescoreQueryWeight(parser.floatValue());
-                } else if ("score_mode".equals(fieldName)) {
-                    String sScoreMode = parser.text();
-                    if ("avg".equals(sScoreMode)) {
-                        rescoreContext.setScoreMode(ScoreMode.Avg);
-                    } else if ("max".equals(sScoreMode)) {
-                        rescoreContext.setScoreMode(ScoreMode.Max);
-                    } else if ("min".equals(sScoreMode)) {
-                        rescoreContext.setScoreMode(ScoreMode.Min);
-                    } else if ("total".equals(sScoreMode)) {
-                        rescoreContext.setScoreMode(ScoreMode.Total);
-                    } else if ("multiply".equals(sScoreMode)) {
-                        rescoreContext.setScoreMode(ScoreMode.Multiply);
-                    } else {
-                        throw new IllegalArgumentException("[rescore] illegal score_mode [" + sScoreMode + "]");
-                    }
-                } else {
-                    throw new IllegalArgumentException("rescore doesn't support [" + fieldName + "]");
-                }
+    private final BaseParser<ParseCtx> rescoreParser = new BaseParser<>();
+
+    public QueryRescorer() {
+        rescoreParser.registerField((p, c) -> {
+            ParsedQuery parsedQuery = c.searchContext.queryParserService().parse(p);
+            c.rescoreContext.setParsedQuery(parsedQuery);
+        }, "rescore_query", Token.START_OBJECT);
+        rescoreParser.registerField((p, c) -> c.rescoreContext.setQueryWeight(p.floatValue()), "query_weight", Token.VALUE_NUMBER, Token.VALUE_STRING);
```

Comment:
Hmmm - maybe move the function to the last parameter? I think it might be easier to read that way.